src/replay_stage.rs

```diff
@@ -270,146 +116,101 @@ impl ReplayStage {
-                        prev_slot = current_slot;
-                        if my_id == leader_id {
-                            // Create new bank for next slot if we are the leader for that slot
-                            bank = Self::create_and_set_working_bank(
-                                next_slot,
-                                &bank_forks,
-                                &old_bank,
+                    // TODO: fork selection
+                    // vote on the latest one for now
+                    votable.sort();
+
+                    if let Some(latest_slot_vote) = votable.last() {
+                        let bank = bank_forks
+                            .read()
+                            .unwrap()
+                            .get(*latest_slot_vote)
+                            .unwrap()
+                            .clone();
+                        let next_slot = *latest_slot_vote + 1;
+                        let next_leader =
+                            LeaderScheduler::default().slot_leader_at(next_slot, &bank);
+                        cluster_info.write().unwrap().set_leader(next_leader);
+
+                        subscriptions.notify_subscribers(&bank);
+
+                        if let Some(ref voting_keypair) = voting_keypair {
+                            let keypair = voting_keypair.as_ref();
+                            let vote = VoteTransaction::new_vote(
+                                keypair,
+                                *latest_slot_vote,
+                                bank.last_id(),
+                                0,
                             );
-                            current_slot = Some(next_slot);
-                            Self::reset_state(
-                                bank.ticks_per_slot(),
-                                next_slot,
-                                &mut max_tick_height_for_slot,
-                                &mut current_blob_index,
-                            );
-                        } else {
-                            current_slot = None;
+                            cluster_info.write().unwrap().push_vote(vote);
                         }
-
-                        if leader_id != current_leader_id {
-                            // TODO: Remove this soon once we boot the leader from ClusterInfo
-                            cluster_info.write().unwrap().set_leader(leader_id);
+                        if next_leader == my_id {
+                            let tpu_bank =
+                                Bank::new_from_parent_and_id(&bank, my_id, next_slot);
+                            bank_forks.write().unwrap().insert(next_slot, tpu_bank);
                         }
-
                         // Always send rotation signal so that other services like
                         // RPC can be made aware of last slot's bank
                         to_leader_sender
                             .send(TvuRotationInfo {
```

Comment:
We could potentially send multiple instances of these signals while an older TPU is still running right?

Reply:
Yea. I don’t think it matters.

Reply:
Yeah I think worst case that older fork just stops making progress, and then we can purge it later